src/solve-commit.cc

```diff
@@ -87,47 +94,85 @@ static TriBool show_problem( Zypper & zypper, const ResolverProblem & prob, Prob
-    // correspond to skip/retry/cancel in that order.
+    // correspond to skip/retry/cancel/detailed information in that order.
     // The answers should be lower case letters.
-    popts.setOptions( numbers.str() + _("s/r/c"), default_reply );
+    popts.setOptions( numbers.str() + _("s/r/c/d"), default_reply );
+    skipAnswer    = solution_count;
+    retryAnswer   = solution_count + 1;
+    cancelAnswer  = solution_count + 2;
+    extInfoAnswer = solution_count + 3;
   }
   else
   {
     default_reply = solution_count;
     // translators: answers for dependency problem solution input prompt:
     // "Choose from above solutions by number or cancel"
     // Translate the letter 'c' to whatever is suitable for your language
-    // and to the same as you translated it in the "s/r/c" string
-    // See the "s/r/c" comment for other details.
+    // and to the same as you translated it in the "s/r/c/d" string
+    // See the "s/r/c/d" comment for other details.
     // One letter string  for translation can be tricky, so in case of problems,
     // please report a bug against zypper at bugzilla.novell.com, we'll try to solve it.
-    popts.setOptions( numbers.str() + _("c"), default_reply );
+    popts.setOptions( numbers.str() + _("c/d"), default_reply );
+    cancelAnswer  = solution_count;
+    extInfoAnswer = solution_count + 1;
   }
 
-  if ( !zypper.config().non_interactive )
-    clear_keyboard_buffer();
-  zypper.out().prompt( PROMPT_DEP_RESOLVE, prompt_text, popts, desc_stm.str() );
-  unsigned reply = get_prompt_reply( zypper, PROMPT_DEP_RESOLVE, popts );
-
-  // retry
-  if ( problem_count > 1 && reply == solution_count + 1 )
-    return true;
-  // cancel (one problem)
-  if ( problem_count == 1 && reply == solution_count )
-    return false;
-  // cancel (more problems)
-  if ( problem_count > 1 && reply == solution_count + 2 )
-    return false;
-  // skip
-  if ( problem_count > 1 && reply == solution_count )
-    return indeterminate; // continue with next problem
+  for ( unsigned i = 0; i < solution_count; i++ )
+    popts.setOptionHelp( i, str::Format( _("Choose solution %1%") ) % (i+1) );
+  if ( skipAnswer != -1 ) {
+    popts.setOptionHelp( skipAnswer,  _("Skip problem and continue.") );
+  }
+  if ( retryAnswer != -1 ) {
+    popts.setOptionHelp( retryAnswer, _("Retry solving immediately.") );
+  }
+  popts.setOptionHelp( cancelAnswer,  _("Choose no solution and cancel.") );
+  popts.setOptionHelp( extInfoAnswer, _("Toggle show detailed conflict information.") );
+
+  bool showExtInfo = false;
+  unsigned reply = default_reply;
+
+  while ( true ) {
+
+    if ( !zypper.config().non_interactive )
+      clear_keyboard_buffer();
+
+    std::ostringstream fulltext;
+    fulltext << desc_stm.str();
+    if ( showExtInfo )
+      fulltext << desc_ext_stm.str();
+    fulltext << solution_stm.str();
+
+    zypper.out().prompt( PROMPT_DEP_RESOLVE, prompt_text, popts, fulltext.str() );
```

Comment:
```
  if ( !zypper.config().non_interactive )
    clear_keyboard_buffer();
```
AFAICS almost all calls to `zypper.out().prompt` are _immediately_ preceded by the above. We should do it here too.

Or add a cleanup Commit? `grep -r -6 '\.prompt\>\|clear_keyboard_buffer' src/` shows that there are just a few cases where  `prompt` is called without `clear_keyboard_buffer`.  I suspect this is not intentional, but the call was simply forgotten. If so, we can move the preceding `clear_keyboard_buffer` call into `prompt` and remove them in the code.

Reply:
prompt() is a pure virtual, so we have 3 options:
A) should we move the clear_keyboard_buffer into each implementation
B) implement the clear_keyboard_buffer in the base and make each impl call it 
C) make a non virtual safePrompt() that calls clear_keyboard_buffer and then the virtual prompt()

Any favourites?

Reply:
Revisiting the code I'm not sure whether this is the right location at all. `Out::prompt` just prints the prompt while `clear_keyboard_buffer` should be done before reading the answer. So `src/utils/prompt.cc:get_prompt_reply:L361` (after opening the terminal) is probably a better place to clear ant typeahead.

Reply:
Moved clear_keyboard_buffer into get_prompt_reply and cleaned up all occurences right before a get_prompt_reply call